### PR TITLE
mgr/dashboard_v2: Implement `get_daemons` for RGW

### PR DESCRIPTION
**Pull request by p-se**
Implement `get_daemons` function of `dashboard` module to `dashboard_v2`.

Signed-off-by: Patrick Nawracay <pnawracay@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import json
+import cherrypy
+
+from ..tools import ApiController, RESTController
+from .. import logger
+from mgr_module import MgrModule, CommandResult
+
+
+@ApiController('rgw')
+class Rgw(RESTController):
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py]: You are not making use of the `RESTController` functionality in your code, so you can just extend from `BaseController`.

**Review by rjfd — CHANGES_REQUESTED** [review]

**Comment by LenzGr** [comment]
Looks like this PR needs rebasing?

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 1 above)
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: Thought about it and you're right!

File: src/pybind/mgr/dashboard_v2/tools.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 import cherrypy
 
 
-def ApiController(path):
+def ApiController(path=''):
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tools.py]: I don't understand what is the purpose of this change. If path is an empty string, then what will be the final URL for the controller?

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    def __init__(self):
+        self.daemon = RgwDaemon()
+
+
+@ApiController()
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py]: Why did you not added the `path` argument to this decorator?

Comment (p-se) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: Because it wouldn't be used. It seems like the `path` is only necessary for the initial controller attached, but not for all nested ones.

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 2 above)
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/tools.py]: That depends. If it's the first controller attached, a path makes sense, for any nested controller, it won't be used anymore.

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 3 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: I still don't understand what you are trying to accomplish with the code in this file.
You have two controller classes declared, and the second one is being instantiated in the constructor of the second one.

To be honest I don't even know what's the behavior of your code. Are you accessing the `RgwDaemon` controller using the `/api/rgw/daemon` URL? If that is the case, you don't need to decorate the `RgwDaemon` class with  `@ApiController`.

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 2 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/tools.py]: @p-na "Nested controllers" don't need to be decorated with `@ApiController`

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                        'status': status,
+                        'url': "{0}/api/rgw/daemon/{1}".format(
+                            self.mgr.url_prefix, service['id'])
+                    }
+
+                    daemons.append(daemon)
+
+        return {'daemons': sorted(daemons, key=lambda k: k['id'])}
+
+    def get(self, service_id):
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py]: I think this can be reduced to the following.
```
    def get(self, svc_id):
        rgw_metadata = {}
        rgw_status = {}

        for daemon in self.list():
            if daemon["id"] != svc_id:
                continue

            rgw_metadata = daemon["metadata"]
            rgw_status = daemon["status"]

        return {
            "rgw_id": svc_id,
            "rgw_metadata": rgw_metadata,
            "rgw_status": rgw_status
        }
```

Comment (votdev) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py]: s/service_id/svc_id/ according to https://github.com/openattic/ceph/blob/wip-mgr-dashboard_v2/src/pybind/mgr/mgr_module.py#L395

File: src/pybind/mgr/dashboard_v2/module.py [hunk 5]
```diff
@@ -46,6 +50,7 @@ class Module(MgrModule):
     def __init__(self, *args, **kwargs):
         super(Module, self).__init__(*args, **kwargs)
         logger.logger = self._logger
+        self.url_prefix = ''
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/module.py]: Consider to transform this into a property, thus it is read-only.

```
@property
def url_prefix(self):
    return self._url_prefix
```

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                        'server': server,
+                        'metadata': metadata,
+                        'status': status,
+                        'url': "{0}/api/rgw/daemon/{1}".format(
+                            self.mgr.url_prefix, service['id'])
+                    }
+
+                    daemons.append(daemon)
+
+        return {'daemons': sorted(daemons, key=lambda k: k['id'])}
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py]: I think ``return sorted(daemons, key=lambda k: k['id'])`` is enough here. We do not need to adapt the structure of the returned data from the origin dashboard.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+from __future__ import absolute_import
+
+import json
+
+from ..tools import ApiController, RESTController
+from .. import logger
+
+
+@ApiController('rgw')
+class Rgw(RESTController):
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py]: I think this API should require an authentication, so please add ``@AuthRequired()``

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 6 above)
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: This is exactly the structure from the original dashboard. But note that I slightly (IMO) improved the structure of the detail view.

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 4 above)
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: No, that won't work, because the key `status` doesn't seem to be there all the time. In fact it wasn't there once since I started this implementation but since the current implementation is basically the port of the `dashboard` module, I kept it purposely so that it works if - in some case, the future or for a reason I don't know yet - the status will be returned.

Comment (p-se) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: If possible, I'd like to keep `service_id`. It may be just my opinion but I like the expressiveness. Also, I wasn't able to translate `svc_id` to `service_id` when I first read it. And it's in the Zen of Python: Explicit is better than implicit. Of course there are different opinions about this and if yours or others are strong ones, let me know why that is.

Comment (LenzGr) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: What's the name that the upstream Ceph project uses here? As we just discussed today, we should not change names on a whim, as it makes it impossible to find all occurrences of a variable across the entire code base.

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 6 above)
Comment (votdev) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: Hmmm, but if i request the daemons via api/deamon i do not expect a return object that is a dict with a key ``daemons``. This is somewhat doubled. I expect an array of deamons that i can immediatelly process.

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 2 above)
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/tools.py]: Revised

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 6 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: If this will return the entire list of daemons, I would prefer it to simply return the list of daemons.

Only in situations where we have server side pagination and filtering, we would need a object as the return value. And I think in oA we return the list on a 'result' property.

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 4 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: IMO I think in this particular case I would keep it as `service_id` since it's more descriptive.

If we were overriding some function of the `MgrModule` or other Ceph module then we should maintain the same signature and thus keep the same parameter names.

Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: Also, we are not changing the meaning of the parameter, we are just using a bit longer form of the same word

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py]: I'd use `svc_id`  here, as the term `service_id` is not really used in ceph. I agree with Lenz here.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/module.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/module.py]: Or
```
@property
def url_prefix(self):
    url_prefix = urljoin('/', self.get_config('url_prefix', default=''))
    return url_prefix.rstrip('/')
```

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Review by sebastian-philipp — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 2 above)
Comment (rjfd) [reply src/pybind/mgr/dashboard_v2/tools.py]: please revert this change, since is not needed anymore

**Review by rjfd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+
+@ApiController('rgw')
+@AuthRequired()
+class Rgw(RESTController):
+
+    def __init__(self):
+        self.daemon = RgwDaemon(self.mgr)
+
+
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/controllers/rgw.py:17]: My experience while implementing another controller is that this 'sub' controller also needs the @AuthRequired() decorator, otherwise this API endpoint is public.

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 8 above)
Comment (p-se) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py:17]: Tested it two times and it's not required.

**Review by sebastian-philipp — APPROVED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/rgw.py (hunk 8 above)
Comment (votdev) [reply src/pybind/mgr/dashboard_v2/controllers/rgw.py:17]: Can't confirm that, but that can be fixed later.

**Review by votdev — APPROVED** [review]

**Review by rjfd — APPROVED** [review]
